### DISCUSSION TITLE
Breaking changes in Rspack v1.0.0

### DISCUSSION THREAD

**Discussion by chenjiahan** [web-infra-dev/rspack#6626] [Core]
Rspack v1.0.0 is scheduled to be released in July 2024.

This thread is used to collect and discuss planned breaking changes for Rspack v1.0.0.

**Comment by chenjiahan** [comment]
## Remove `rspackExperiments.emotion`

Remove `rspackExperiments.emotion`, as it can be replaced by [@swc/plugin-emotion](https://www.npmjs.com/package/@swc/plugin-emotion).

- Install:

```bash
npm i @swc/plugin-emotion -D
```

- Configure:

```diff
module.exports = {
  module: {
    rules: [
      {
        test: /\.jsx$/,
        loader: "builtin:swc-loader",
        options: {
-         rspackExperiments: {
-            emotion: true,
-         },
          jsc: {
+           experimental: {
+             plugins: [["@swc/plugin-emotion", {}]],
+           },
          },
        },
      },
    ],
  },
};
```

**Reply by chenjiahan** [reply]
Example: https://github.com/rspack-contrib/rspack-examples/pull/86/files

**Reply by h-a-n-a** [reply]
Honestly, I was thinking of the status quo of wasm plugin. Exposing the wasm plugin is necessarily exposing the stability of internal SWC ast...

**Reply by chenjiahan** [reply]
Agreed, this is an issue that requires attention.

The good news is that SWC plugins like emotion/relay/styledComponents are maintained by the SWC team, so when users use the latest version of Rspack and these SWC plugins, they are expected to work properly.

**Reply by chenjiahan** [reply]
Related PR:
- https://github.com/web-infra-dev/rspack/pull/6862
- https://github.com/web-infra-dev/rspack/pull/6892

**Comment by hardfist** [comment]
## experiments.css

Switch `experiments.css` to false by default to align with webpack @JSerFeng

**Reply by chenjiahan** [reply]
https://github.com/web-infra-dev/rspack/pull/6910

**Comment by hardfist** [comment]
## ModuleConcatenationPlugin

Enable `ModuleConcatenationPlugin` by default in production to reduce bundle size and align with webpack @JSerFeng

**Comment by chenjiahan** [comment]
## Remove `rspackExperiments.relay`

Remove `rspackExperiments.relay`, as it can be replaced by [@swc/plugin-relay](https://www.npmjs.com/package/@swc/plugin-relay).

- Install:

```bash
npm i @swc/plugin-relay -D
```

- Configure:

```diff
module.exports = {
  module: {
    rules: [
      {
        test: /\.jsx$/,
        loader: "builtin:swc-loader",
        options: {
-         rspackExperiments: {
-           relay: true,
-         },
          jsc: {
+           experimental: {
+             plugins: [["@swc/plugin-relay", {}]],
+           },
          },
        },
      },
    ],
  },
};
```

**Reply by chenjiahan** [reply]
Related PR:
- https://github.com/web-infra-dev/rspack/pull/6862
- https://github.com/web-infra-dev/rspack/pull/6892

**Comment by chenjiahan** [comment]
## Remove `rspackExperiments.styledComponents`

Remove `rspackExperiments.styledComponents`, as it can be replaced by [@swc/plugin-styled-components](https://www.npmjs.com/package/@swc/plugin-styled-components).

- Install:

```bash
npm i @swc/plugin-styled-components -D
```

- Configure:

```diff
module.exports = {
  module: {
    rules: [
      {
        test: /\.jsx$/,
        loader: "builtin:swc-loader",
        options: {
-         rspackExperiments: {
-           styledComponents: {},
-         },
          jsc: {
+           experimental: {
+             plugins: [["@swc/plugin-styled-components", {}]],
+           },
          },
        },
      },
    ],
  },
};
```

**Reply by chenjiahan** [reply]
Related PR:
- https://github.com/web-infra-dev/rspack/pull/6862
- https://github.com/web-infra-dev/rspack/pull/6892

**Comment by chenjiahan** [comment]
## Integrate lightningcss

- **Step 1**: Introduce lightningcss as an optional CSS minimizer.
- **Step 2**: If it works well, make lightningcss as the default CSS minimizer and remove `swc_css`.
- **Step 3**: Consider providing a `builtin:lightningcss-loader` like `builtin:swc-loader`, so users can use lightningcss to transform CSS.

Ref: 

- https://github.com/vercel/next.js/discussions/46536
- https://github.com/vercel/next.js/pull/58471
- https://github.com/vercel/turbo/pull/6456

**Reply by chenjiahan** [reply]
@ahabhgk cc~

**Comment by ahabhgk** [comment]
Remove [`resolve.tsConfigPath`](https://www.rspack.dev/zh/config/resolve#resolvetsconfigpath), prefer `resolve.tsConfig = string` or `resolve.tsConfig.configPath = string` instead

**Reply by ahabhgk** [reply]
There are some resolve questions still unclear to me:

- Do we need to support `resolve.plugins`?
  - If we support `resolve.plugins`, should we move tsConfig as a plugin instead of an options?
- Do we need to use native resolver in js side for some loader (css-loader, sass-loader, ...)? or keep it as is

**Reply by hardfist** [reply]
1. I'm not sure if we implement resolve hook, whether it can implement all the functionalities reoslve.plugins provide? if so, I think support resolve hook is enough for us
2. even we don't support resolve.plugins, we can still move resolve.tsConfig as native plugins(but not as resolver plugins)
3. if the native-resolver can meets the need of css-loader sass-loader and some common used commnity loaders, use native-resolver seems better for alignments between js resolver and css|sass resolver

**Reply by ahabhgk** [reply]
1. I don't think it's an easy thing to implement `resolve.plugins` before v1.0, this even need to refactor lots of code in oxc-resolver since enhanced-resolve is plugins composed together based on the plugin system, to align every detail for `resolve.plugins` is very hard.
2. In most cases the only usage for `resolve.plugins` is supporting `tsconfig`, IMO it's ok to keep it as an options in v1.0
3. I think https://github.com/web-infra-dev/rspack/pull/4945 is good enough, most loader are not using those incompatible api, only using `resolver.resolve`, the use case of those incompatible api is `resolve.plugins` I guess

**Reply by chenjiahan** [reply]
I would like to add some usage scenarios of `resolve.plugins`:

- [ModuleScopePlugin of create-react-app](https://github.com/facebook/create-react-app/blob/0a827f69ab0d2ee3871ba9b71350031d8a81b7ae/packages/react-dev-utils/ModuleScopePlugin.js#L14)
- [pnp-webpack-plugin](https://github.com/arcanis/pnp-webpack-plugin)

**Reply by ceopaludetto** [reply]
Unplugin also relies on resolve plugins

- [Resolve plugin of unplugin (webpack version)](https://github.com/unjs/unplugin/blob/a01da85ed2f9924e8361df14a21e887e14bf0e67/src/webpack/index.ts#L70)
- https://github.com/unjs/unplugin/issues/311

**Reply by chenjiahan** [reply]
Related PR: https://github.com/web-infra-dev/rspack/pull/6872

**Comment by colinaaa** [comment]
Change the default value of [`optimization. chunkIds`](https://webpack.js.org/configuration/optimization/#optimizationchunkids) and [`optimization.moduleIds`](https://webpack.js.org/configuration/optimization/#optimizationmoduleids) to `'natural'` when `mode === 'none'`. This is the same as [webpack](https://github.com/webpack/webpack/blob/d9ce3b1f87e63c809d8a19bbd92257d65922e81f/lib/config/defaults.js#L1395-L1404). See: #6685 and #6692.

**Reply by LingyuCoder** [reply]
https://github.com/web-infra-dev/rspack/pull/6956

**Comment by h-a-n-a** [comment]
JavaScript APIs

**Reply by h-a-n-a** [reply]
Removed Compilation.currentNormalModuleHooks and mark most of the fields as readonly: https://github.com/web-infra-dev/rspack/pull/6859

**Reply by h-a-n-a** [reply]
Changed `Compilation.errors` and `Compilation.warnings` API from `StatsError` to `JsRspackError`: https://github.com/web-infra-dev/rspack/pull/6900

**Comment by h-a-n-a** [comment]
SWC Wasm plugin breaking https://github.com/web-infra-dev/rspack/pull/6887

**Comment by LingyuCoder** [comment]
Remove stats.modules[].profile.integration, it related to persistent cache which rspack has different design with webpack https://github.com/web-infra-dev/rspack/pull/6947

**Comment by h-a-n-a** [comment]
Remove fields from `SwcJsMinimizerPlugin` 
https://github.com/web-infra-dev/rspack/pull/6950.
https://github.com/web-infra-dev/rspack/pull/6970.
Default value remains untouched.

**Comment by LingyuCoder** [comment]
use `@swc/plugin-prefresh` instead of the builtin one https://github.com/web-infra-dev/rspack/pull/7069

**Comment by abenhamdine** [comment]
public environnement variables must now be defined, even if empty, see https://github.com/web-infra-dev/rsbuild/issues/3029#issuecomment-2253653007

**Comment by ahabhgk** [comment]
Refactor minimizer plugin options, see https://github.com/web-infra-dev/rspack/pull/7444
